### ISSUE TITLE
If server killed in a moment after table metadata was changed in ZooKeeper but before it was changed locally (by ALTER MODIFY COLUMN) query, it may refuse to start.

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#26327]
https://clickhouse-test-reports.s3.yandex.net/0/1b4d2cbc6da217f0f89ba791f40fec3ca05d464e/stress_test_(thread).html

```
4095002:2021.07.07 14:51:03.075351 [ 42181 ] {} <Information> test_21.concurrent_mutate_kill (4b4fbb74-5d6c-44b0-8b4f-bb745d6c44b0): Metadata changed in ZooKeeper. Applying changes locally.
...
SIGKILL received
...
2021.07.07 14:58:20.636554 [ 355147 ] {} <Error> Application: DB::Exception: Table columns structure in ZooKeeper is different from local table structure: Cannot attach table `test_21`.`concurrent_mutate_kill` from metadata file /var/lib/clickhouse/store/c2a/c2a4cf77-eba8-42ed-82a4-cf77eba862ed/concurrent_mutate_kill.sql from query ATTACH TABLE test_21.concurrent_mutate_kill UUID '4b4fbb74-5d6c-44b0-8b4f-bb745d6c44b0' (`key` UInt64, `value` UInt64) ENGINE = ReplicatedMergeTree('/clickhouse/tables/01593_concurrent_alter_mutations_kill_test_21/concurrent_mutate_kill', '1') PARTITION BY key % 100 ORDER BY key SETTINGS max_replicated_mutations_in_queue = 1000, number_of_free_entries_in_pool_to_execute_mutation = 0, max_replicated_merges_in_queue = 1000, index_granularity = 8192: while loading database `test_21` from path /var/lib/clickhouse/metadata/test_21
```

**Comment by PedroTadim** [comment]
🕵️ 

Still reproducible on 26.10. Two things worth adding to the original report.

**A per-column `SETTINGS` clause hits this too.** The 2021 report was a plain type change. This variant differs by a single clause, which makes it easy to misread as corruption rather than as a metadata desync:

```
Local columns:
`c1` Nullable(DateTime64(3))

Zookeeper columns:
`c1` Nullable(DateTime64(3))	SETTINGS (min_compress_block_size = 512)
```

That is, `ALTER TABLE t MODIFY COLUMN c1 Nullable(DateTime64(3)) SETTINGS (min_compress_block_size = 512)` followed by the server going down before the local metadata is rewritten. Every other column matches byte for byte.

**The window is still explicit in `StorageReplicatedMergeTree::executeMetadataAlter`.** As of 26.10 it branches on the database engine:

```cpp
bool is_in_replicated_database = database->getEngineName() == "Replicated";

if (is_in_replicated_database)
{
    ...
    for (auto & op : requests)
        alter_txn->addOp(std::move(op));
    requests.clear();
    /// Requests will be executed by database in setTableStructure
}
else
{
    zookeeper->multi(requests, /* check_session_valid */ true);
}

LOG_INFO(log, "Metadata changed in ZooKeeper. Applying changes locally.");
...
setTableStructure(table_id, alter_context, std::move(columns_from_entry), metadata_diff, entry.alter_version);
```

Under `Replicated` the ZooKeeper ops are folded into the `ZooKeeperMetadataTransaction` and committed together with `setTableStructure`. Under `Atomic` / `Ordinary` they are committed first and the local update follows, so any termination between the two leaves ZooKeeper ahead and `checkTableStructure` refuses the table on the next start. The `Metadata changed in ZooKeeper. Applying changes locally.` line quoted in the original report is exactly that boundary — a log showing it with no following `Applied changes to the metadata of the table` is the signature.

Worth stating explicitly, because it shapes the fix: this asymmetry is not an oversight. `Replicated` can be atomic only because that database's own metadata already lives in Keeper, so one transaction covers both writes. For `Atomic` / `Ordinary` the local side is a `.sql` file on local disk, and there is no transaction spanning Keeper and the local filesystem. Reordering the two writes does not help either — `checkTableStructure` tests for equality, so local-ahead-of-Keeper is rejected just as local-behind is.

**The consequence is broader than it was in 2021.** With async table loading the failure no longer stops at one table:

```
Code: 122 INCOMPATIBLE_COLUMNS
  -> Code: 695 Load job 'load table db.t' failed
  -> Code: 696 Load job 'startup table db.t'
  -> Code: 722 ASYNC_LOAD_WAIT_FAILED
```

Any later query that waits on that load job fails with 722, including queries against *other* tables in the same database, so one desynced table can make a whole database unusable rather than just failing to attach.

Which leaves catching up at startup as the only general fix — essentially what @kafka1991 suggested above, now that `metadata_version` is tracked per replica. When the local metadata version is behind the one in ZooKeeper, apply the ZooKeeper metadata through the existing `setTableStructure` path instead of throwing `INCOMPATIBLE_COLUMNS`. `checkTableStructure` already takes a `metadata_version` out-parameter and a `strict_check` flag, but the non-strict mode only suppresses the check rather than reconciling, so the reconciliation step would be new.
